include/oneapi/tbb/collaborative_once_flag.h

```diff
@@ -0,0 +1,126 @@
+
+#ifndef __TBB_collaborative_once_flag_H
+#define __TBB_collaborative_once_flag_H
+
+#include "tbb/task_arena.h"
+#include "tbb/task_group.h"
+#include "tbb/task.h"
+
+#include <atomic>
+#include <memory>
+#include <mutex>
+
+namespace tbb {
+
+namespace detail {
+
+class run_once {
+    //! A tbb::empty_task allows provides something for other threads to wait on.
+    tbb::task_arena my_arena;
+    detail::d1::wait_context my_wait_context;
+    mutable task_group_context my_context; ///< mutable so we can get the status in a const.
+
+public:
+    run_once()
+        : my_arena(task_arena::attach{})
+        , my_wait_context(0)
+        , my_context(task_group_context::bound,
+            task_group_context::default_traits | task_group_context::concurrent_wait) {}
+
+    ~run_once() {
+        // Caller is responsable for not pulling this out from under someone else.
+        // We could use a std::shared_mutex though...
+    }
+
+    [[nodiscard]] bool is_canceling() const { return my_context.is_group_execution_cancelled(); }
+
+    template <typename F>
+    task_group_status run_and_wait(F&& f) {
+        detail::d1::function_stack_task<F> t{ f, my_wait_context };
+        my_wait_context.reserve();
+        bool cancellation_status = false;
+        auto func = [&]() {
+            detail::try_call([&] { execute_and_wait(t, my_context, my_wait_context, my_context); }).on_completion([&] {
+                cancellation_status = my_context.is_group_execution_cancelled();
+                // Don't bother resetting the context. This isn't reusable.
+            });
+        };
+        my_arena.execute([&func]() { return tbb::this_task_arena::isolate(func); });
+        return cancellation_status ? canceled : complete;
+    }
+
+    /*! A call to wait() potentially participates in the work of run_and_wait().
+     * If run_and_wait() throws or is canceled, this returns canceled.
+     */
+    task_group_status wait() {
+        bool cancellation_status = false;
+        auto f = [&]() {
+            detail::try_call([&] {
+                detail::d1::wait(my_wait_context, my_context);
+            }).on_completion([&] {
+                cancellation_status = my_context.is_group_execution_cancelled();
```

Comment:
It seems we do not need all cancellation logic in `wait`. Additionally, consider the comment about `exception` vs `cancellation`.